### ISSUE TITLE
(Feature request) Implement MCP (Model Context Protocol) 

### ISSUE THREAD

**Issue by dwdcth** [TriliumNext/trilium#5302]
### Describe feature

I hope to implement the MCP protocol, so that on Claude's desktop version, I can directly use Claude to retrieve notes, generate responses, or save conversations directly to notes. So TriliumNext will be an AI note.

### Additional Information

The mcp [site](https://github.com/modelcontextprotocol) ,   and [document](https://modelcontextprotocol.io/).

**Comment by tan-yong-sheng** [comment]
Hi guys, I am enhancing the search capabilities of triliumnext-mcp using Trilium Search DSL (refer: https://triliumnext.github.io/Docs/Wiki/search.html) - just partial support, please see more examples at https://github.com/tan-yong-sheng/triliumnext-mcp/blob/main/docs/search-query-examples.md.

Basically, I add a few more functions:
- `list_child_note`:  List direct children of a parent note (like Unix ls command)
- `list_descendant_note`: List ALL descendant notes recursively in database or subtree (like Unix find command)
- `search` function now support queries like "get me the notes created or modified in past 3 months", "Search "n8n" notes created since 2020, ordered by creation date descending, limit 10" and much more. see more at [MCP Search Examples & Usage Guide (Advanced Params)](https://github.com/tan-yong-sheng/triliumnext-mcp/blob/main/docs/search-query-examples.md)
- `append_note`: Add content below the existing note content, which is append-only operation. This prevent modification to the existing note content.
- `update_note` which tends to overwrite the content will create revision by default before update operation unless user explicitly mention not to. It is to prevent data loss of the note.

- Introduction of the environment variable: PERMISSIONS, which you set to either READ, WRITE, or READ;WRITE, to control the level of access your triliumnext-mcp instance has to the notes.
  - READ: The MCP can only fetch and view notes.
  - WRITE: The MCP can create or modify notes, but not read existing ones.
  - READ;WRITE: Full access, allowing both reading and writing operations.

  This gives you fine-grained control over what the MCP can do inside your Trilium environment, making it easier to balance security and functionality depending on your use case.

## Reference to this pull request:
- https://github.com/tan-yong-sheng/triliumnext-mcp/pull/9
- https://github.com/tan-yong-sheng/triliumnext-mcp/pull/8

---

Actually at my dev branch, I am trying to implement more support for Trilium Search DSL: https://github.com/tan-yong-sheng/triliumnext-mcp/blob/dev/docs/search-query-examples.md, yet it is still under development... Basically, I let LLM to input JSON parameters required to generate the Trilium Search DSL, and ya, there are a lot of manual rules to be defined, and it's hard to make it 1-to-1 coverage for all Trilium Search DSL... Welcome for contribution, thanks.

**Comment by tan-yong-sheng** [comment]
UPDATED: I've just released an beta version that could be installed via `npx -y triliumnext-mcp@beta`.

- for `search_notes` function, added more coverage to trilium search DSL such as searching by label, relations (untested), and support for flexible 'AND' and 'OR' logic if multiple conditions needed to be achieved in single search query, for example, search notes created OR modified in past 7 days.... And much more at: https://github.com/tan-yong-sheng/triliumnext-mcp/blob/dev/docs/search-query-examples.md... also, remove list_child_notes and list_descendant_notes as their functionality is added to search_notes 
- The resolve_note_id tool converts folder names (e.g., ‘n8n Templates’) into precise note IDs by matching exact or partial titles, ensuring accurate folder resolution for subsequent operations.


<img width="1517" height="853" alt="Image" src="https://github.com/user-attachments/assets/9e8ab63f-a083-4672-af57-251d4b766ca9" />
For the above image, it supports both partial/exact matching to the title of the note folder.

<img width="1527" height="728" alt="Image" src="https://github.com/user-attachments/assets/2353979c-cb99-47d7-88a3-de1424452b9f" />

The new beta release is here:
- https://www.npmjs.com/package/triliumnext-mcp/v/0.3.0-beta.1


Read more here before you test: https://github.com/tan-yong-sheng/triliumnext-mcp/tree/dev

**Comment by eliandoran** [comment]
On our side we don't plan to implement MCP as we'd prefer a different approach for LLMs (a way that's more directly integrated in the application itself).

@tan-yong-sheng 's solution should suffice https://github.com/tan-yong-sheng/triliumnext-mcp.

**Comment by raffaelemancuso** [comment]
> On our side we don't plan to implement MCP as we'd prefer a different approach for LLMs (a way that's more directly integrated in the application itself).
> 
> [@tan-yong-sheng](https://github.com/tan-yong-sheng) 's solution should suffice https://github.com/tan-yong-sheng/triliumnext-mcp.

I think that's different application.

Integration the LLM in the application itself allow you to generate LLM response within Trilium.

An MCP server would allow to save the response generated by a third-party LLM into Trilium (without copying and pasting of course)

**Comment by marvil6** [comment]
Agreed 100%. My Trilium tree is extremely large, being able to spin up an agent on OpenWebUI or similar would be extremely useful.

**Comment by eliandoran** [comment]
@marvil6 , @raffaelemancuso , Trilium already comes with an MCP server which allows external applications to interact with the Trilium DB. see https://docs.triliumnotes.org/user-guide/llm#built-in-mcp.
